crates/typst/src/geom/gradient.rs

```diff
@@ -0,0 +1,1070 @@
+    /// between 0% and 100%) or an [angle]($angle). Any value outside
+    /// of this range will be clamped.
+    ///
+    /// 🚧 The angle will be mainly used for conic gradients once they are ready.
+    #[func]
+    pub fn sample(
+        &self,
+
+        /// The position at which to sample the gradient.
+        t: RatioOrAngle,
+    ) -> Color {
+        let value: f64 = t.into();
+
+        match self {
+            Self::Linear(linear) => sample_stops(&linear.stops, linear.space, value),
+        }
+    }
+
+    /// Sample the gradient at the given positions.
+    ///
+    /// The positions are either the progress along the gradient (a [ratio]($ratio)
+    /// between 0% and 100%) or an [angle]($angle). Any value outside
+    /// of this range will be clamped.
+    ///
+    /// 🚧 Angles will be mainly used for conic gradients once they are ready.
+    #[func]
+    pub fn samples(
+        &self,
+
+        /// The positions at which to sample the gradient.
+        #[variadic]
+        ts: Vec<RatioOrAngle>,
+    ) -> Array {
+        Array::from(
+            ts.into_iter()
+                .map(|t| self.sample(t).into_value())
+                .collect::<EcoVec<_>>(),
+        )
+    }
+
+    /// Create a sharp version of this gradient.
+    ///
+    /// This is particularly useful for creating color lists that come from
+    /// a preset gradient.
+    ///
+    /// ```example
+    /// #let grad = gradient.linear(..gradient.rainbow(20))
+    /// #rect(width: 100pt, height: 20pt, fill: grad)
+    /// #rect(width: 100pt, height: 20pt, fill: grad.sharp(5))
+    /// ```
+    #[func]
+    pub fn sharp(
+        &self,
+
+        /// The number of stops in the gradient.
+        n: Spanned<usize>,
+
+        /// How much to smooth the gradient.
+        #[default(Spanned::new(Ratio::zero(), Span::detached()))]
+        #[named]
+        smoothness: Spanned<Ratio>,
+    ) -> SourceResult<Gradient> {
+        if n.v < 2 {
+            bail!(n.span, "sharp gradients must have at least two stops");
+        }
+
+        if smoothness.v.get() < 0.0 || smoothness.v.get() > 1.0 {
+            bail!(smoothness.span, "smoothness must be between 0 and 1");
+        }
+
+        let smoothness = smoothness.v.get();
+        let colors = (0..n.v)
+            .flat_map(|i| {
+                let c = self
+                    .sample(RatioOrAngle::Ratio(Ratio::new(i as f64 / (n.v - 1) as f64)));
+
+                [c, c]
+            })
+            .collect::<Vec<_>>();
+
+        let mut positions = Vec::with_capacity(n.v * 2);
+        let p = |i| i as f64 * 1.0 / n.v as f64;
+
+        let t = smoothness * 1.0 / (4.0 * n.v as f64);
+        for i in 0..n.v {
+            let mut j = 2 * i;
+            positions.push(p(i));
+            if j > 0 {
+                positions[j] += t;
+            }
+
+            j += 1;
+            positions.push(p(i + 1));
+            if j < colors.len() - 1 {
+                positions[j] -= t;
+            }
```

Comment:
Perhaps those variables could have a bit more meaningful names? (I know this is mostly maths:tm:, but maybe we can turn these into names which are a bit more helpful to future maintainers, or maybe document what they mean?)

Reply:
Truth is, I wrote this code months ago and even I struggle to understand it, as far as `p`, `t`, `i`, and `j`:
- `p` and `t` I can rename
- `i` and `j` are very common variables for indices